Add macOS pkg installer to deployment (#7554)

paulober opened the pull request:
- Fixes #7554

paulober committed:
Add macOS pkg installer to deployment

paulober commented:
See the connected issue for more details on the implementation

paulober commented:
@vilmibm @williammartin Any updates on this?

paulober committed:
Merge branch 'trunk' into feature-macos-pkg-installer

paulober committed:
Fix typo in macOS pkg installer build script

williammartin commented:
@paulober Sorry that your PR is languishing in review purgatory, when I went on parental leave I asked @vilmibm to take ownership but he has since left GitHub. I have not forgotten about it and we are currently figuring out how to deal with our PR backlog with team churn.

paulober commented:
@samcoe Why has the "discuss" label be removed? Does GH not plan on giving the macOS community an easier installer?

paulober commented:
Will my work now finally after nearly one year be accepted and merged, or should i just delete it and distribute a gh pkg for macOS on my own?

_(You are currently maintaining 12 artifacts for linux with 3 different packages per one of the 4 supported architectures.
And macOS people still have to install it manually from a zip folder or rely on a heavy third-party tool.)_

paulober committed:
Merge branch 'trunk' into feature-macos-pkg-installer

paulober commented:
@andyfeller Hi, if you find some time may you review this PR for merging please?

andyfeller commented:
> @andyfeller Hi, if you find some time may you review this PR for merging please?

@paulober : 🫡 Firstly, let me say thank you for putting the time and effort into this as well as your patience given the impact attrition has taken on the team over the past year.

I'm blocking off time first thing tomorrow morning to review this.  Not having a background with `.pkg` packaging, I also want to compare this to efforts in GitHub Desktop and Mobile areas to compare as a colleague of mine has been working at refactoring how we build and sign within the Mac ecosystem.

andyfeller commented on build/macOS/distribution.xml:
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8" standalone="yes"?>
+<installer-gui-script minSpecVersion="1">
+    <title>Github Cli</title>
```
I suspect this might be the way it is due to the Developer ID we use but wanted to make sure:

```suggestion
    <title>GitHub CLI</title>
```

paulober replied:
No, that text is not dependent on the developer id. Just the title shown in the installer UI.

andyfeller commented on build/macOS/distribution.xml:
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8" standalone="yes"?>
+<installer-gui-script minSpecVersion="1">
```
For documentation on this file, see [Apple Distribution Definition reference](https://developer.apple.com/library/archive/documentation/DeveloperTools/Reference/DistributionDefinitionRef/Chapters/Introduction.html)

paulober replied:
Right, i think we can change the minSpecVersion to `2`. But i'll check that first.

andyfeller replied:
Just to be clear, I'm mostly adding reference comments because completely unfamiliar with the spec and I use comments like sticky notes for myself 😅

paulober replied:
oh, sorry. didn't know that.

andyfeller commented on build/macOS/distribution.xml:
```diff
@@ -0,0 +1,19 @@
+    <title>Github Cli</title>
+    <license file="LICENSE" mime-type="text/plain"/>
+    <options hostArchitectures="arm64,x86_64" customize="never" require-scripts="false" allow-external-scripts="false"/>
+    <domains enable_localSystem="true"/>
```
**Should GitHub CLI be system-wide, per-user, or can be anywhere?**

[GitHub Desktop](https://desktop.github.com/) zips up the application and on start up will ask the user if they want to move it from Downloads _(in my case)_ to Applications.

<img width="1072" alt="Screenshot of GitHub Desktop startup, asking if user wants to relocate it to Applications directory" src="https://github.com/cli/cli/assets/2089743/822998f7-9a9b-4d05-8e71-7fd28817c90f">

[Reference](https://developer.apple.com/library/archive/documentation/DeveloperTools/Reference/DistributionDefinitionRef/Chapters/Distribution_XML_Ref.html#//apple_ref/doc/uid/TP40005370-CH100-SW35)

andyfeller replied:
@williammartin : would appreciate your 2 cents here as `brew` installs packages such that any user on the workstation can add `/opt/homebrew` to their path where the configuration is per user.

williammartin replied:
I think I'd be inclined to say `/usr/local/bin`. When I look at the contents of my `/usr/local/bin` it feels like a familiar idea. Homebrew seems to have a variety of reasons relating to [Apple Silicon to use `/opt/homebrew`](https://github.com/Homebrew/brew/issues/9177).

I would anticipate the vast majority of users:
 * Don't feel the need to use amd64 and arm versions of `gh`
 * Have a single user account on their personal machine

So I think I would suggest we sidestep that mess and install into `/usr/local/bin`, knowing that it's on the path already.

What do you think?

andyfeller commented on build/macOS/distribution.xml:
```diff
@@ -0,0 +1,19 @@
+    <choices-outline>
+        <line choice="gh-cli"/>
+    </choices-outline>
+    <choice id="gh-cli" title="Github Cli (universal)">
```
```suggestion
    <choice id="gh-cli" title="GitHub CLI (universal)">
```

andyfeller commented on build/macOS/distribution.xml:
```diff
@@ -0,0 +1,19 @@
+    <pkg-ref id="com.github.cli.pkg" auth="Root" packageIdentifier="com.github.cli">#com.github.cli.pkg</pkg-ref>
+    <pkg-ref id="com.github.cli.pkg" auth="Root" packageIdentifier="com.github.cli">
+        <bundle-version/>
+    </pkg-ref>
```
**What is the impact of this being empty versus being managed from package data?**

[Reference](https://developer.apple.com/library/archive/documentation/DeveloperTools/Reference/DistributionDefinitionRef/Chapters/Distribution_XML_Ref.html#//apple_ref/doc/uid/TP40005370-CH100-SW62)

> Defines the version of the bundles delivered by the parent element. You do not typically specify this element; `productbuild` inserts it from the actual package data when the product archive is created.

paulober replied:
I just checked the code i wrote one year ago and removed some of this xml as it's not needed. 
The pkg-ref bundle-version part is generated by productbuild and it also uses this empty bundle-version as the real version can be looked up by `installer` in the included pkg or the generated named version property on pkg-ref.

andyfeller commented on script/pkgmacos:
```diff
@@ -0,0 +1,108 @@
+#!/bin/zsh
+set -e
+
+print_help() {
+    cat <<EOF
+To build a universal pkg for macOS:
+  script/pkgmacos <tag-name>
+
+To build and sign set APPLE_DEVELOPER_INSTALLER_ID environment variable before.
+For example, if you have a signing identity with the identifier 
+"Developer ID Installer: Your Name (ABC123DEF)" set it in the variable.
+EOF
+}
```
@paulober : Is this something you've written wholesale or is this based on any existing OSS package script we should reference?

paulober replied:
I wrote this but i looked up some of your style in the other build files (like the print_help) while learning your code-base. I did mention the developer environment variable after i found out how the build stuff for macOS targets work to make it easier for others to use this.

andyfeller replied:
Thanks for the insight, just to be clear this was pure curiosity!  I am always learning by looking at variations on the same approach, like listening to covers of a specific song to appreciate nuances.

andyfeller commented on script/pkgmacos:
```diff
@@ -0,0 +1,108 @@
+    for file in ./share/man/man1/gh*.1; do
+        cp "$file" "${payload_man1}"
+    done
+    # Include only Zsh completions,
+    # the recommended/only option on macOS since Catalina for default shell.
+    cp "./share/zsh/site-functions/_gh" "${payload_zsh_site_functions}"
```
🙉 🙊 🙈

paulober replied:
I this is the part i might replace with the gh completion approach you mentioned.

andyfeller replied:
heh I was mostly avoiding the rabbit hole around Mac OS, BSD, licensing challenges, and past experiences

andyfeller commented on script/pkgmacos:
```diff
@@ -0,0 +1,108 @@
+    # merge binaries
+    merge_binaries
+
+    # build pkg
+    pkgbuild \
+    --root "$payload_root" \
+    --identifier "com.github.cli" \
+    --version "$tag_name" \
+    --install-location "/" \
+    "dist/com.github.cli.pkg"
```
Additional indentation to improve readability

```suggestion
    # build pkg
    pkgbuild \
        --root "$payload_root" \
        --identifier "com.github.cli" \
        --version "$tag_name" \
        --install-location "/" \
        "dist/com.github.cli.pkg"
```

andyfeller commented on script/pkgmacos:
```diff
@@ -0,0 +1,108 @@
+
+    # build distribution
+    if [ -n "$APPLE_DEVELOPER_INSTALLER_ID" ]; then
+        # build and sign production package with license
+        productbuild \
+        --distribution "./build/macOS/distribution.xml" \
+        --resources "./build/macOS/resources" \
+        --package-path "./dist" \
+        --timestamp \
+        --sign "${APPLE_DEVELOPER_INSTALLER_ID?}" \
+        "./dist/gh_${tag_name}_macOS_universal.pkg"
```
Additional indentation to improve readability

```suggestion
        # build and sign production package with license
        productbuild \
            --distribution "./build/macOS/distribution.xml" \
            --resources "./build/macOS/resources" \
            --package-path "./dist" \
            --timestamp \
            --sign "${APPLE_DEVELOPER_INSTALLER_ID?}" \
            "./dist/gh_${tag_name}_macOS_universal.pkg"
```

andyfeller replied:
**How do you feel about building an argument array for invoking `productbuild` to reduce redundancy?**

The only difference between both of these blocks is whether the `--sign ...` argument is passed.  I mention the idea of an argument array that is used for the conditional elements to reduce errors from modifying one code block and not the other.

```shell
    PRODUCTBUILD_ARGS=()

    # include signing if developer installer ID is present
    if "${APPLE_DEVELOPER_INSTALLER_ID}"; then
        PRODUCTBUILD_ARGS+=(--sign "${APPLE_DEVELOPER_INSTALLER_ID}")
    fi

    # build package with license
    productbuild \
        "${PRODUCTBUILD_ARGS[@]}" \
        --distribution "./build/macOS/distribution.xml" \
        --resources "./build/macOS/resources" \
        --package-path "./dist" \
        --timestamp \
        "./dist/gh_${tag_name}_macOS_universal.pkg"
```

paulober replied:
Sounds good. I'll integrate it.

andyfeller commented on script/pkgmacos:
```diff
@@ -0,0 +1,108 @@
+    else
+        echo "skipping macOS pkg code-signing; APPLE_DEVELOPER_INSTALLER_ID not set" >&2
+
+        # build production package with license without signing
+        productbuild \
+        --distribution "./build/macOS/distribution.xml" \
+        --resources "./build/macOS/resources" \
+        --package-path "./dist" \
+        "./dist/gh_${tag_name}_macOS_universal.pkg"
```
Additional indentation to improve readability

```suggestion
        # build production package with license without signing
        productbuild \
            --distribution "./build/macOS/distribution.xml" \
            --resources "./build/macOS/resources" \
            --package-path "./dist" \
            "./dist/gh_${tag_name}_macOS_universal.pkg"
```

andyfeller commented on script/pkgmacos:
```diff
@@ -0,0 +1,108 @@
+    cp "LICENSE" "build/macOS/resources"
+    touch .gi
+
+    # build distribution
```
For information on productbuild, see [Apple "Packaging Mac software for distribution"](https://developer.apple.com/documentation/xcode/packaging-mac-software-for-distribution#Build-an-Installer-package)

andyfeller commented on script/pkgmacos:
```diff
@@ -0,0 +1,108 @@
+        "./dist/gh_${tag_name}_macOS_universal.pkg"
+    fi
+
+    # remove temp installer so it does not get uploaded
+    rm "dist/com.github.cli.pkg"
```
Is it worth setting up a trap to ensure this is removed regardless?

paulober replied:
Depending if this script failing file stop the upload step. But to be save why not. This ensures that the temporary file does not persist if the script is interrupted or fails at any point.

andyfeller commented on script/pkgmacos:
```diff
@@ -0,0 +1,108 @@
+
+tag_name=""
+
+while [ $# -gt 0 ]; do
+  case "$1" in
+  -h | --help )
+    print_help
+    exit 0
+    ;;
+  -* )
+    printf "unrecognized flag: %s\n" "$1" >&2
+    exit 1
+    ;;
+  * )
+    tag_name="$1"
+    shift 1
+    ;;
+  esac
+done
```
**Are there any conditions we should short circuit running this script?**

- Running it on unsupported OS
- If `pkgbuild` or `productbuild` aren't installed
- ...

paulober replied:
Not really required but it's better.
I adopted your suggested requirements and as unsupported OS i set all macOS versions prior to macOS 12 (Monterey).
For Linux it's not accounted for as it would exit with code 127 at the start anyways as `sw_vers` is not available on Linux.

paulober replied:
Do you think that's sufficient for this build script or should i also write a safe fail for Linux?

andyfeller commented on .goreleaser.yml:
```diff
@@ -10,7 +10,7 @@ before:
       {{ if eq .Runtime.Goos "windows" }}echo{{ end }} make manpages GH_VERSION={{.Version}}
     - >-
-      {{ if ne .Runtime.Goos "linux" }}echo{{ end }} make completions
+      {{ if ne .Runtime.Goos "windows" }}echo{{ end }} make completions
```
Could you explain this change?

paulober replied:
It was just for testing as i never worked with goreleaser before. I'll remove it.

andyfeller reviewed (commented):
@paulober : again thanks for your patience especially as my background in packaging is predominantly OSS with RPMs and DEBs.  

few minor nits and some honest questions I'd appreciate your thoughts on while running down hubbers on GitHub Desktop and Mobile to understand these changes in context to those product lines.  additionally, I still want to ensure the CLI team has an opportunity to review these.

paulober commented:
> > @andyfeller Hi, if you find some time may you review this PR for merging please?
> 
> @paulober : 🫡 Firstly, let me say thank you for putting the time and effort into this as well as your patience given the impact attrition has taken on the team over the past year.
> 
> I'm blocking off time first thing tomorrow morning to review this. Not having a background with DMG packaging, I also want to compare this to efforts in GitHub Desktop and Mobile areas to compare as a colleague of mine has been working at refactoring how we build and sign within the Mac ecosystem.

Thanks for taking the time to review. I'll address all comments.
Refactoring the macOS target build and signing process sounds like a good project, looking forward to seeing the result.

paulober committed:
Update pkg title

Co-authored-by: Andy Feller <andyfeller@github.com>

andyfeller commented:
The only product tangentially relevant to GitHub that built `.pkg` was [Mac OS Git installer](https://github.com/timcharper/git_osx_installer/blob/83b50f1923364eb491336dfce74375f65a599498/Makefile#L177-L180).

paulober committed:
Update choice title

Co-authored-by: Andy Feller <andyfeller@github.com>

paulober committed:
Indentation fix

Co-authored-by: Andy Feller <andyfeller@github.com>

paulober committed:
Fix indentation

Co-authored-by: Andy Feller <andyfeller@github.com>

paulober committed:
Fix indentation

Co-authored-by: Andy Feller <andyfeller@github.com>

paulober committed:
Undo goreleaser change

Signed-off-by: paulober <44974737+paulober@users.noreply.github.com>

paulober committed:
Fix indentation

Signed-off-by: paulober <44974737+paulober@users.noreply.github.com>

paulober committed:
Removed redundant specifications

Signed-off-by: paulober <44974737+paulober@users.noreply.github.com>

paulober committed:
Cleanup pkgmacos build script

Signed-off-by: paulober <44974737+paulober@users.noreply.github.com>

paulober commented:
Does gh cli has a minimum supported macOS version? If so, a script could be added to the installer to check for minimum macOS version.

paulober committed:
Change minimum build script macOS version

Signed-off-by: paulober <44974737+paulober@users.noreply.github.com>

andyfeller commented:
> Does gh cli has a minimum supported macOS version? If so, a script could be added to the installer to check for minimum macOS version.

If we look at the current [`gh` homebrew formula](https://github.com/Homebrew/homebrew-core/blob/ba2e0401002c5d7963b7927922eb89e1bfaf3633/Formula/g/gh.rb#L15-L22), Monterey is the oldest of current versions supported now.  If we [go back to 1.0.0 it was High Sierra](https://github.com/Homebrew/homebrew-core/commit/557638387d9253faaba59b4976f2a1e6ba0776f1).

As we're not backporting older versions, I assume to error on the side of [latest supported Mac OS versions](https://endoflife.date/macos)

andyfeller commented on .github/workflows/deployment.yml:
```diff
@@ -108,6 +108,19 @@ jobs:
         run: |
           shopt -s failglob
           script/sign dist/gh_*_macOS_*.zip
+      - name: Build universal macOS pkg installer
+        if: inputs.environment != 'production'
+        env:
+          TAG_NAME: ${{ inputs.tag_name }}
+        run: script/pkgmacos "$TAG_NAME"
```
**How have you been testing `script/pkgmacos` locally?**

I was wondering how large the new `.pkg` files were, so tried building it locally.  I thought about trying to build this locally only to realize we probably want to add a new target to `Makefile` as this involves multiple steps.

Here's how far I've got so far before realizing I should simply ask 😅 

```shell
andyfeller@Andys-MBP:cli/cli ‹feature-macos-pkg-installer›$ ./script/pkgmacos 
To build a universal pkg for macOS:
  script/pkgmacos <tag-name>

To build and sign set APPLE_DEVELOPER_INSTALLER_ID environment variable before.
For example, if you have a signing identity with the identifier 
"Developer ID Installer: Your Name (ABC123DEF)" set it in the variable.
andyfeller@Andys-MBP:cli/cli ‹feature-macos-pkg-installer›$ ./script/pkgmacos v0.0.0
build_pkg:7: no matches found: ./share/man/man1/gh*.1
andyfeller@Andys-MBP:cli/cli ‹feature-macos-pkg-installer›$ vim Makefile 
andyfeller@Andys-MBP:cli/cli ‹feature-macos-pkg-installer›$ make
go build -trimpath -ldflags "-X github.com/cli/cli/v2/internal/build.Date=2024-05-23 -X github.com/cli/cli/v2/internal/build.Version=v2.49.2-50-g9454d5e7 " -o bin/gh ./cmd/gh
andyfeller@Andys-MBP:cli/cli ‹feature-macos-pkg-installer›$ ./script/pkgmacos v0.0.0
build_pkg:7: no matches found: ./share/man/man1/gh*.1
andyfeller@Andys-MBP:cli/cli ‹feature-macos-pkg-installer›$ vim Makefile 
andyfeller@Andys-MBP:cli/cli ‹feature-macos-pkg-installer›$ make manpages
go run ./cmd/gen-docs --man-page --doc-path ./share/man/man1/
andyfeller@Andys-MBP:cli/cli ‹feature-macos-pkg-installer›$ ./script/pkgmacos v0.0.0
```

paulober replied:
Concerning the pkg size, its about 24MB:
`-rw-r--r--@ 1 paulober  staff    24M May 24 14:10 gh_v2.99.8_macOS_universal.pkg`

To test it locally:
```zsh
make manpages
make completions
./script/release --local "v2.99.8" --platform macos
./script/pkgmacos v2.99.8
```

And for the Makefile target. Maybe i can read into it. But as of now i know nothing about the syntax within a Makefile (only cmake).

paulober replied:
I take that back, it wasn't that difficult. You can now build a pkg using `make macospkg VERSION=v2.99.8`. (goreleaser required)

andyfeller replied:
this is really going to help the team, thank you for saving us future time 🫶

paulober committed:
Fix directory already exists

Signed-off-by: paulober <44974737+paulober@users.noreply.github.com>

paulober committed:
Added make macospkg target

Signed-off-by: paulober <44974737+paulober@users.noreply.github.com>

paulober committed:
Fix distribution.xml + min macos version requirements

Signed-off-by: paulober <44974737+paulober@users.noreply.github.com>

paulober committed:
Added native min os version blocking

Signed-off-by: paulober <44974737+paulober@users.noreply.github.com>

paulober committed:
Integrate argument array to remove duplicate code

Signed-off-by: paulober <44974737+paulober@users.noreply.github.com>

paulober commented:
I just wanted to let you know that I have now finished all the suggested changes as well as my own changes that I wanted to incorporate.

paulober committed:
Merge branch 'trunk' into feature-macos-pkg-installer

andyfeller commented:
> I just wanted to let you know that I have now finished all the suggested changes as well as my own changes that I wanted to incorporate.

thanks for the high sign ✋  just to set expectations for you, I'm going to devote time this afternoon to review these before extended Memorial Day weekend.  Being back on Wednesday, I want you to know this is something I've committed internally to wrap this up by next week at the latest.

andyfeller commented on Makefile:
```diff
@@ -93,3 +93,11 @@ uninstall:
 	rm -f ${DESTDIR}${datadir}/bash-completion/completions/gh
 	rm -f ${DESTDIR}${datadir}/fish/vendor_completions.d/gh.fish
 	rm -f ${DESTDIR}${datadir}/zsh/site-functions/_gh
+
+.PHONY: macospkg
+macospkg: manpages completions
+ifndef VERSION
+	$(error VERSION is not set. Use `make macospkg VERSION=vX.Y.Z`)
+endif
+	./script/release --local "$(VERSION)" --platform macos
+	./script/pkgmacos $(VERSION)
```
Checking with and without `VERSION` ❤️ 

```shell
andyfeller@Andys-MBP:cli/cli ‹feature-macos-pkg-installer›$ make macospkg
go run ./cmd/gen-docs --man-page --doc-path ./share/man/man1/
go build -trimpath -ldflags "-X github.com/cli/cli/v2/internal/build.Date=2024-05-24 -X github.com/cli/cli/v2/internal/build.Version=v2.49.2-64-g4db87793 " -o bin/gh ./cmd/gh
mkdir -p ./share/bash-completion/completions ./share/fish/vendor_completions.d ./share/zsh/site-functions
bin/gh completion -s bash > ./share/bash-completion/completions/gh
bin/gh completion -s fish > ./share/fish/vendor_completions.d/gh.fish
bin/gh completion -s zsh > ./share/zsh/site-functions/_gh
Makefile:100: *** VERSION is not set. Use `make macospkg VERSION=vX.Y.Z`.  Stop.
andyfeller@Andys-MBP:cli/cli ‹feature-macos-pkg-installer›$ make macospkg VERSION=v0.0.0
go run ./cmd/gen-docs --man-page --doc-path ./share/man/man1/
build.go: `bin/gh` is up to date.
mkdir -p ./share/bash-completion/completions ./share/fish/vendor_completions.d ./share/zsh/site-functions
bin/gh completion -s bash > ./share/bash-completion/completions/gh
bin/gh completion -s fish > ./share/fish/vendor_completions.d/gh.fish
bin/gh completion -s zsh > ./share/zsh/site-functions/_gh
./script/release --local "v0.0.0" --platform macos
goreleaser release -f .goreleaser.generated.yml --clean --skip-validate --skip-publish --release-notes=$TMPDIR/tmp.6ZAch4nuYr
  • starting release...
  • loading                                          path=.goreleaser.generated.yml
  • DEPRECATED: --skip-publish was deprecated in favor of --skip=publish, check https://goreleaser.com/deprecations#-skip for more details
  • DEPRECATED: --skip-validate was deprecated in favor of --skip=validate, check https://goreleaser.com/deprecations#-skip for more details
  • skipping announce, publish and validate...
  • loading environment variables
  • getting and validating git state
    • couldn't find any tags before "v0.0.0"
    • git state                                      commit=4db87793cd2315a18a33a9b56e6234b594b8b16a branch=feature-macos-pkg-installer current_tag=v0.0.0 previous_tag=<unknown> dirty=false
    • pipe skipped                                   reason=validation is disabled
  • parsing tag
  • setting defaults
    • DEPRECATED:  archives.rlcp  should not be used anymore, check https://goreleaser.com/deprecations#archivesrlcp for more info
    • DEPRECATED:  archives.rlcp  should not be used anymore, check https://goreleaser.com/deprecations#archivesrlcp for more info
    • DEPRECATED:  archives.rlcp  should not be used anymore, check https://goreleaser.com/deprecations#archivesrlcp for more info
  • running before hooks
    • running                                        hook= make manpages GH_VERSION=0.0.0
    • running                                        hook=echo make completions
    • took: 2s
  • checking distribution directory
    • cleaning dist
  • loading go mod information
  • build prerequisites
  • writing effective config file
    • writing                                        config=dist/config.yaml
  • building binaries
    • building                                       binary=dist/macos_darwin_arm64/bin/gh
    • building                                       binary=dist/macos_darwin_amd64_v1/bin/gh
    • running hook                                   hook=./script/sign '/Users/andyfeller/Documents/workspace/cli/cli/dist/macos_darwin_arm64/bin/gh'
skipping macOS code-signing; APPLE_DEVELOPER_ID not set
    • running hook                                   hook=./script/sign '/Users/andyfeller/Documents/workspace/cli/cli/dist/macos_darwin_amd64_v1/bin/gh'
skipping macOS code-signing; APPLE_DEVELOPER_ID not set
    • took: 9s
  • generating changelog
    • loaded "/var/folders/y5/q89b76cj7ws55x3trcyllf_80000gn/T/tmp.6ZAch4nuYr", but it is empty
  • archives
    • creating                                       archive=dist/gh_0.0.0_macOS_amd64.zip
    • creating                                       archive=dist/gh_0.0.0_macOS_arm64.zip
    • took: 3s
  • calculating checksums
  • storing release metadata
    • writing                                        file=dist/artifacts.json
    • writing                                        file=dist/metadata.json
  • you are using deprecated options, check the output above for details
  • release succeeded after 14s
  • thanks for using goreleaser!
./script/pkgmacos v0.0.0
pkgbuild: Inferring bundle components from contents of pkg_payload
pkgbuild: Wrote package to ./dist/com.github.cli.pkg
skipping macOS pkg code-signing; APPLE_DEVELOPER_INSTALLER_ID not set
productbuild: Wrote product to ./dist/gh_v0.0.0_macOS_universal.pkg
```

andyfeller committed:
Remove `v` prefix when `pkgmacos` is called

Existing Mac OS release artifacts use the tag name / version in the file name but drop the `v` prefix.  This does the same for the Mac OS installer.

andyfeller commented on script/pkgmacos:
```diff
@@ -0,0 +1,129 @@
+    exit 1
+    ;;
+  * )
+    tag_name="${1#v}"
```
@paulober : just a heads up, I pushed up a small change to strip the leading `v` so the resulting artifact matched the others 👍 

```shell
andyfeller@Andys-MBP:cli/cli ‹feature-macos-pkg-installer*›$ ll dist 
total 103056
-rw-r--r--  1 andyfeller  staff   1.2K May 24 15:09 artifacts.json
-rw-r--r--  1 andyfeller  staff   4.0K May 24 15:09 config.yaml
-rw-r--r--  1 andyfeller  staff   182B May 24 15:09 gh_0.0.0_checksums.txt
-rw-r--r--  1 andyfeller  staff    13M May 24 15:09 gh_0.0.0_macOS_amd64.zip
-rw-r--r--  1 andyfeller  staff    12M May 24 15:09 gh_0.0.0_macOS_arm64.zip
-rw-r--r--  1 andyfeller  staff    24M May 24 15:09 gh_0.0.0_macOS_universal.pkg
drwxr-xr-x  3 andyfeller  staff    96B May 24 15:09 macos_darwin_amd64_v1
drwxr-xr-x  3 andyfeller  staff    96B May 24 15:09 macos_darwin_arm64
-rw-r--r--  1 andyfeller  staff   211B May 24 15:09 metadata.json
```

andyfeller commented:
**Demo of using local build from `make macospkg`**

https://github.com/cli/cli/assets/2089743/6cdef88f-ba89-443f-bcfb-b6a393378a69

andyfeller committed:
Update readme about MacOS pkg

andyfeller reviewed (approved):
I think this is good to go with our next release being the ultimate test of any problems.

paulober commented:
Thanks for taking the time to bring the PR of the finish line. Lets see how it works with the next release. I also though about maybe adding a second install script that removes old gh man pages in case some get removed with a new release they wouldn't remain on the users disk if they upgrade with a new pkg...

williammartin commented:
Unfortunately this broke the release: https://github.com/cli/cli/actions/runs/9272921562/job/25511828490

Looking into it now but if you have any idea that would be great, thanks.

williammartin commented:
I see in https://github.com/cli/cli/pull/7555#discussion_r1612259525 the example command and the new makefile target build `manpages` and `completions` before building the release and running the packaging script. I would guess that the completions are not available in our `macos` job and therefore cannot be found. Still investigating.

williammartin commented:
And I think we don't have completions because they are only build on `linux`: https://github.com/cli/cli/blob/f1dedc9abbf0aeb8e5090eb791e3a31d2c2f1711/.goreleaser.yml#L13

I think they are only built on linux because they are only packaged up inside the `.deb` and `.rpm` files: https://github.com/cli/cli/blob/f1dedc9abbf0aeb8e5090eb791e3a31d2c2f1711/.goreleaser.yml#L93-L98

Perhaps this is as simple as changing that conditional so that they are also produced on `macos` though it would be a bit of spooky action at a distance that we're expecting to nab the completions at some future point in the release script. Happy to try it for now, leave a comment and revisit.

williammartin commented:
Created https://github.com/cli/cli/pull/9136 which I believe fixes this.

paulober commented:
Oh right, i guess that's why i originally added: https://github.com/cli/cli/pull/7555#discussion_r1608467603

paulober commented:
Sorry, shouldn't have removed it.

williammartin commented:
No worries, Workflows mostly being collections of cobbled together scripts makes things extremely challenging to write and review. I'm not at all surprised we missed something.

williammartin commented:
It looks like we built the pkg successfully but we failed to sign it: https://github.com/cli/cli/actions/runs/9284093254/job/25545720127#step:9:11

Looking at the workflow we expect something called `APPLE_DEVELOPER_INSTALLER_ID` to be set: https://github.com/cli/cli/blob/3620e791dc1c6db79ca2f6aaeb66fb93a745d93a/.github/workflows/deployment.yml#L122 but we don't have any variable with that name.

Is it supposed to be `vars.APPLE_DEVELOPER_ID`?

paulober commented:
No, because to sign installers you need a special purpose certificate from Apple. In the portal where you also created your APPLE_DEVELOPER_ID certificate you can create a new one for installer signing. It should start with `Developer ID Installer ...`

williammartin commented:
Thanks. Because Apple enforce a limit of 5 certificates, the CLI team does not own or manage them.  If this is a different certificate as you say, I don't think it's likely this will be resolved today. We may have to either revert this or release without it being signed in the interim. Will look into this now but I suspect the people with the answer are in the US timezone.

paulober commented:
Ok understand. Not sure what you mean with a limit of 5 certificate. I have about 15 active. Alternatively i could offer to sign the first release manually with my Developer ID Installer certificate or we release it unsigned (maybe with a little note in the release).

williammartin commented:
Well, the limit is documented [here](https://developer.apple.com/help/account/create-certificates/create-developer-id-certificates) and that is what has been communicated to our team.

I think it would be better to release unsigned than signed by your personal cert, thanks. Unsigned may be acceptable as an MVP that we improve. The binaries themselves weren't [signed from the beginning either](https://github.com/cli/cli/issues/6970). I'll discuss with @andyfeller when he comes online.

paulober commented:
> Well, the limit is documented [here](https://developer.apple.com/help/account/create-certificates/create-developer-id-certificates) and that is what has been communicated to our team.

Oh right, didn't know that.

williammartin commented:
And as you can see from https://github.blog/2023-01-30-action-needed-for-github-desktop-and-atom-users/, we really, really want to take care of these limited certs 😬

williammartin commented:
@paulober when you sign your binaries, how do you ensure that the certificate is in the keychain and accessible. For example, my understanding of our release process is that we [import the certificate used for signing](https://github.com/cli/cli/blob/3620e791dc1c6db79ca2f6aaeb66fb93a745d93a/.github/workflows/deployment.yml#L91-L92) and then later when we [`codesign`](https://github.com/cli/cli/blob/3620e791dc1c6db79ca2f6aaeb66fb93a745d93a/script/sign#L41), `codesign` checks the keychain and uses that cert for signing.

I believe the same is true for `productbuild` based on https://thegreyblog.blogspot.com/2014/06/os-x-creating-packages-from-command_2.html "Signing the Product Archive" section. Perhaps the certificate is the same, but then I would also expect we need to allow `productbuild` access to it on both these lines: https://github.com/cli/cli/blob/3620e791dc1c6db79ca2f6aaeb66fb93a745d93a/.github/workflows/deployment.yml#L91-L92

paulober commented:
If I sign from terminal manually i normally install the certificate via drag and drop. (not applicable here)
Yes, `productbuild` and `codesign` both pull the certificate from the security keychain .

It should not be the same as the Developer ID Installer and Developer ID Application are different certificates with different "permissions" (though i never tried signing a binary with a Developer ID Installer certificate).
And yes for signing without sudo, `-T /usr/bin/productbuild` needs to be added to the `security import` command.
Concerning the second command I think we don't have to change stuff there and i'm not sure if `codesign:` is even a valid value here

williammartin commented:
Thanks. I think we're in shared understanding. I also believe we need another cert and we need to load it into the keychain.

> Concerning the second command I think we don't have to change stuff there and i'm not sure if codesign: is even a valid value here

Yeh I'm not sure about that either. The docs for `set-key-partition-list` say:

> The "partition list" is an extra parameter in the ACL which limits access to the key based on an application's code signature. You must present the keychain's password to change a partition list. If you'd like to run /usr/bin/codesign with the key, "apple:" must be an element of the partition list.

Which doesn't indicate the need for `codesign`, nor can I find mention of people doing that in other places: https://stackoverflow.com/questions/39868578/security-codesign-in-sierra-keychain-ignores-access-control-settings-and-ui-p

andyfeller commented:
> Thanks. I think we're in shared understanding. I also believe we need another cert and we need to load it into the keychain.

Agreed 😞 

Relevant docs: [Apple "Packaging Mac Software for distribution > Build an Installer package"](https://developer.apple.com/documentation/xcode/packaging-mac-software-for-distribution#Build-an-Installer-package):

> If you choose to distribute your product in an Installer package, start by determining your Installer signing identity. Choose the right identity for your distribution channel:
> 
> - If you’re distributing an app on the Mac App Store, use a Mac Installer Distribution signing identity. This is named 3rd Party Mac Developer Installer: <TeamID>, where <TeamID> identifies your team.
> 
> - If you’re distributing a product independently, use a Developer ID Installer-signing identity. This is named Developer ID Installer: <TeamID>, where <TeamID> identifies your team.

Additional doc on [Creating Developer ID certificates](https://developer.apple.com/help/account/create-certificates/create-developer-id-certificates/)

williammartin commented:
> The "partition list" is an extra parameter in the ACL which limits access to the key based on an application's code signature. You must present the keychain's password to change a partition list. If you'd like to run /usr/bin/codesign with the key, "apple:" must be an element of the partition list.


Some answers in here indicate that `codesign` should be part of the partition list:+ https://stackoverflow.com/a/40870033

williammartin commented:
@paulober when you say:

> Ok understand. Not sure what you mean with a limit of 5 certificate. I have about 15 active

When getting certificates now we noticed on the Apple developer website and the [docs](https://developer.apple.com/documentation/xcode/packaging-mac-software-for-distribution) there was a difference between `Mac Installer Distribution signing identity` and `Developer ID Installer-signing identity`. The former can be used for distributing on the App Store and it didn't seem to have limitations and the latter required us to be the account owner and seemed much more restricted. Perhaps that is the difference in our use cases?

paulober commented:
> Some answers in here indicate that `codesign` should be part of the partition list:+ https://stackoverflow.com/a/40870033

Strange, i'm sill looking for a documentation about where to find these partitions and which are required for which usecase/tools.

paulober commented:
I'm not sure if i understand you question correctly. But I guess you where asking about the differences between these two certificates in our case:
So the Developer ID Installer named certificates have a different purpose in their metadata (`1.2.840.113635.100.4.13`) whereas the Developer ID Application named one have `Code Signing ( 1.3.6.1.5.5.7.3.3 )` as their purpose. So if you try to sign a `pkg` with `productsign --sign "Developer ID Application ..."` you'll get an error: `An installer signing identity (not an application signing identity) is required for signing flat-style products.` And i guess Gatekeeper also don't like it.

So you would need to create a csr and upload it to apple dev center and request a Developer ID Installer which you can then inject like the code-signing certificate into the action.

williammartin commented:
> I'm not sure if i understand you question correctly. But I guess you where asking about the differences between these two certificates in our case

Ah not quite. I was referring to the fact there are two different identities that can be used to sign `.pkgs`. One is `Mac Installer Distribution signing identity` which has no limit to the number you can create, and seems to be used to sign pkgs for the Mac store, and the other is `Developer ID Installer-signing identity` which seems to be limited to 5. I just wondered if maybe you had 15 of the first one, whereas we needed the second since we are distributing outside the mac store.

From the [docs](https://developer.apple.com/documentation/xcode/packaging-mac-software-for-distribution):

>If you choose to distribute your product in an Installer package, start by determining your Installer signing identity. Choose the right identity for your distribution channel:
>
>If you’re distributing an app on the Mac App Store, use a Mac Installer Distribution signing identity. This is named 3rd Party Mac Developer Installer: <TeamID>, where <TeamID> identifies your team.
>
>If you’re distributing a product independently, use a Developer ID Installer-signing identity. This is named Developer ID Installer: <TeamID>, where <TeamID> identifies your team.

These are both separate from `Developer ID Application identity` which we currently use to sign the binaries, and we are in agreement that they cannot be used to sign pkgs.

paulober commented:
Oh, yes, I might overlooked that one. So for my understanding the Installer Distribution one is not that restricted because Apple can take a look at the software before it is distributed to the end user. Whereas a Developer ID Installer certificate gives you much higher responsibility. 

And by "about 15 certificates" I meant the amount of all certificates across all types as i though you meant this limit.
